src/main/kotlin/org/koil/org/OrganizationService.kt

```diff
@@ -0,0 +1,149 @@
+    private val userService: UserService,
+    private val accountRepository: AccountRepository,
+) : OrganizationService {
+
+    companion object {
+        private val LOGGER: Logger = LoggerFactory.getLogger(AdminServiceImpl::class.java)
+    }
+
+    override fun setupOrganization(request: OrganizationSetupRequest): OrganizationCreatedResult {
+        return try {
+            createOrganizationAndUser(request)
+        } catch (exception: Exception) {
+            LOGGER.error("Failed to setup organization and admin user with exception [$exception]", exception)
+            OrganizationCreatedResult.CreationFailed
+        }
+    }
+
+    override fun getAllOrganizations(): List<Organization> {
+        return organizationRepository.findAll().toList()
+    }
+
+    @Transactional
+    fun createOrganizationAndUser(request: OrganizationSetupRequest): OrganizationCreatedResult {
+        val organization = organizationRepository.save(request.toOrganization())
+
+        return when (val adminUserCreationResult = userService.createUser(request.toUserCreationRequest(organization.signupLink))) {
+            is UserCreationResult.CreatedUser -> OrganizationCreatedResult.CreatedOrganization(
+                organization,
+                adminUserCreationResult.account
+            )
+            UserCreationResult.InvalidSignupLink -> OrganizationCreatedResult.UserCreationFailed(adminUserCreationResult)
+            UserCreationResult.UserAlreadyExists -> OrganizationCreatedResult.UserCreationFailed(adminUserCreationResult)
+        }
+    }
+
+    override fun getOrganizationDetails(queryingAsAccount: Long): Organization {
+        val account = getAccount(queryingAsAccount)
+        validateOrgOwnerStatus(account)
+        val organization = organizationRepository.findOrganizationByOrganizationId(account.organizationId)
+        require(organization != null) {
+            "The organization was not found for this account. That should be impossible- every account must have a valid organization linked to it."
+        }
+        return organization
+    }
+
+    @Transactional
+    override fun updateAccount(
+        requestor: Long,
+        userToUpdate: Long,
+        request: UpdateAccountRequest
+    ): OrgAccountUpdateResult {
+        val requestorAccount = getAccount(requestor)
+
+        require(requestorAccount.isAdmin() || requestorAccount.isOrganizationOwner()) {
+            "Attempting to update an account as a non-admin user!"
+        }
+
+        val account: Account = accountRepository.findByIdOrNull(userToUpdate)
+            ?: return OrgAccountUpdateResult.CouldNotFindAccount
+
+        if (!requestorAccount.isAdmin()) {
+            require(account.organizationId == requestorAccount.organizationId) {
+                "Attempting to update an account for an organization the user does not belong to."
+            }
+            require(!request.authorities.contains(UserAuthority.ADMIN)) {
```

Comment:
Similarly, can we add a method to the request for something like `containsOnlyAllowedAuthorities`?

In this way, if we add another restricted authority in the future (something like `UserAuthority.READ_ADMIN`), then we only have to modify the request class rather than the places where it's used.